### ISSUE TITLE
Bug trying to dispose videoplayer controller

### ISSUE THREAD

**Issue by Kypsis** [jonataslaw/getx#368]
**Fill in the template. Issues that do not respect the model will be closed.**

**Describe the bug**
Navigating back to a previous screen after disposing video player controller results in a red screen: 
```
════════ Exception caught by widgets library ═══════════════════════════════════
The following assertion was thrown building Overlay-[LabeledGlobalKey<OverlayState>#d0b60](state: OverlayState#5c02d(entries: [OverlayEntry#286c0(opaque: true; maintainState: false), OverlayEntry#0ac2e(opaque: false; maintainState: true), OverlayEntry#8f714(opaque: true; maintainState: false), OverlayEntry#f915b(opaque: false; maintainState: true)])):
A VideoPlayerController was used after being disposed.

Once you have called dispose() on a VideoPlayerController, it can no longer be used.
The relevant error-causing widget was
    GetMaterialApp 
lib\main.dart:18
When the exception was thrown, this was the stack
#0      ChangeNotifier._debugAssertNotDisposed.<anonymous closure> 
package:flutter/…/foundation/change_notifier.dart:106
#1      ChangeNotifier._debugAssertNotDisposed 
package:flutter/…/foundation/change_notifier.dart:112
#2      ChangeNotifier.removeListener 
package:flutter/…/foundation/change_notifier.dart:167
#3      _VideoPlayerState.deactivate 
package:video_player/video_player.dart:545
#4      StatefulElement.deactivate 
package:flutter/…/widgets/framework.dart:4722
...
════════════════════════════════════════════════════════════════════════════════
I/flutter (29617): [BACK ROUTE] /challenge
I/flutter (29617): [GET] onClose of ChallengeController called
I/flutter (29617): [GET] ChallengeController deleted from memory

════════ Exception caught by widgets library ═══════════════════════════════════
Duplicate GlobalKeys detected in widget tree.
``` 
**To Reproduce**
1. Go to participation screen, wait for video to load
2. Go back to previous screen
3. See error

**Expected behavior**
Navigating back to the previous screen from a screen with video player should not raise an exception.

**Screenshots**

**Flutter Version:**
1.17.5

**Get Version:**
3.2.2

**Describe on which device you found the bug:**
Android Pixel 2 emulator API 28

**Minimal reproduce code**
participation_controller.dart:
```dart
import 'package:get/get.dart';
import 'package:video_player/video_player.dart';

import 'package:challenge_me_app/data/repositories/database_repository.dart';
import 'package:challenge_me_app/data/models/video.dart';

class ParticipationController extends GetxController {
  static ParticipationController get to => Get.find();

  Video video;

  @override
  void onInit() async {
    try {
      video = await DatabaseRepository().getVideo(Get.arguments);

      VideoPlayerController controller =
          VideoPlayerController.network(video.url);
      await controller.initialize();

      controller.setLooping(true);
      video.controller = controller;
      update();

      await video.controller.play();
    } catch (error) {
      print(error);
    }
  }

  @override
  void onClose() {
      super.onClose();
      video.controller.dispose();
  }
}
```

participation_screen.dart:
```dart
import 'package:flutter/material.dart';
import 'package:get/get.dart';

import 'package:challenge_me_app/ui/widgets/video_player_widget/video_player_widget.dart';
import 'package:challenge_me_app/ui/widgets/video_player_widget/video_progress_widget.dart';
import './participation_controller.dart';

class ParticipationScreen extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: GetBuilder<ParticipationController>(
        init: ParticipationController(),
        /* global: false, */
        builder: (c) {
          return c.video != null && c.video.controller.value.initialized
              ? Stack(
                  children: [
                    VideoPlayerWidget(controller: c.video.controller),
                    VideoProgressWidget(controller: c.video.controller)
                  ],
                )
              : Container();
        },
      ),
    );
  }
}
```

I've tried using the _global: false_ flag but then the Get controller doesn't get disposed and video player controller keeps playing in the background. How can I dispose of the video player controller correctly?

**Comment by Kypsis** [comment]
Calling the dispose after a time seems to work but that seems very error prone approach to the problem:
```
@override
  void onClose() {
    try {
      super.onClose();
      Timer(Duration(milliseconds: 500), () {
        if (video.controller != null) {
          video.controller.dispose();
          video.controller = null;
        }
      });
    } catch (error) {
      print(error);
    }
  }
```

**Comment by jonataslaw** [comment]
I was unable to reproduce your error locally, but I'm thinking of putting the onClose run to after the stack. Just to make sure that inserting onClose after the animation will solve this problem for you, could you test this?

```dart
 @override
  void onClose() {
    try {
      Future.delayed(Duration.zero, () {
        if (video.controller != null) {
          video.controller.dispose();
          video.controller = null;
        }
      });
    } catch (error) {
      print(error);
    }
  }
```

If this works for you I will change the onClose to be called after the transition animation.

**Comment by Kypsis** [comment]
Sadly that does not work with Duration.zero. Minimal duration that works is Duration(milliseconds: 351). So a process that takes 350 milliseconds conflicts with something?

**Comment by Kypsis** [comment]
Narrowing it down more, it definitely has a connection with transition duration. It will crash if the dispose is called earlier than transitionDuration + some variable time under 50 ms eg  with Future.delayed(Duration(milliseconds: 351), the transitionDuration that will not crash seems to vary between 309 and 330.
```
GetMaterialApp(    
      transitionDuration: Duration(milliseconds: 309),
...
```

**Comment by jonataslaw** [comment]
Do you only use this Controller on this screen?

**Comment by Kypsis** [comment]
The screen that leads to this is:
challenge_screen.dart:
```dart
import 'package:flutter/material.dart';
import 'package:get/get.dart';
import 'package:cached_network_image/cached_network_image.dart';

import 'package:challenge_me_app/constants.dart';
import './challenge_controller.dart';

class ChallengeScreen extends StatelessWidget {
  final ChallengeController c = Get.put(ChallengeController());

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Colors.white,
        iconTheme: IconThemeData(
          color: Colors.black,
        ),
        centerTitle: true,
        elevation: 1,
        title: Text(
          "Challenge page",
          style: TextStyle(color: Colors.black),
        ),
      ),
      body: GetBuilder<ChallengeController>(
        builder: (c) {
          return Container(
            child: c.challenge != null
                ? GridView.builder(
                    gridDelegate: SliverGridDelegateWithFixedCrossAxisCount(
                        crossAxisCount: 3,
                        crossAxisSpacing: 3,
                        mainAxisSpacing: 3),
                    itemCount: c.challenge.length,
                    itemBuilder: (context, index) => Card(
                      child: Container(
                        height: 300.0,
                        child: GestureDetector(
                          onTap: () => Get.toNamed("/participation",
                              arguments: c.challenge[index].id),
                          child: Stack(
                            alignment: Alignment.center,
                            fit: StackFit.expand,
                            children: <Widget>[
                              CachedNetworkImage(
                                imageUrl: "https://placeholder_image_from_somewhere",
                                fit: BoxFit.cover,
                                placeholder: (context, url) => Center(
                                  child: CircularProgressIndicator(),
                                ),
                                errorWidget: (context, url, error) =>
                                    Icon(Icons.error),
                              ),
                              Padding(
                                padding: EdgeInsets.all(8.0),
                                child: Column(
                                  crossAxisAlignment: CrossAxisAlignment.start,
                                  mainAxisAlignment: MainAxisAlignment.end,
                                  children: <Widget>[
                                    Text("Id: ${c.challenge[index].id}",
                                        style: kCardText),
                                    Text(
                                      "User: ${c.challenge[index].user}",
                                      style: kCardText,
                                    ),
                                    Text(
                                      "Created at: ${c.challenge[index].created_at}",
                                      style: kCardText,
                                    ),
                                  ],
                                ),
                              ),
                            ],
                          ),
                        ),
                      ),
                    ),
                  )
                : Container(
                    child: Center(
                      child: CircularProgressIndicator(),
                    ),
                  ),
          );
        },
      ),
    );
  }
}
```

challenge_controller.dart:
```dart
import 'package:get/get.dart';

import 'package:challenge_me_app/data/repositories/database_repository.dart';

class ChallengeController extends GetxController {
  static ChallengeController get to => Get.find();

  List challenge;

  onInit() async {
    challenge = await DatabaseRepository().getChallenge(Get.arguments);
    update();
  }
}
```

**Comment by jonataslaw** [comment]
```dart
GetBuilder<ParticipationController>(
        init: ParticipationController(),
```
I noticed that line of code, and you're not using Bindings.
For your structure, perhaps insert:
smartManagement: SmartManagement.onlyBuilder should match your structure better, try this

**Comment by Kypsis** [comment]
Actually I do use bindings in the main.dart file so this one didn't sadly work. The choice of a video controller in the binding was due to the need to have it be available when app starts and to persist the video controller state between screens. Eg I go to different screen the video pauses and resumes when I come back.
```dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:get/get.dart';

import 'package:challenge_me_app/routes.dart';
import 'package:challenge_me_app/data/shared/auth_controller.dart';
import 'package:challenge_me_app/data/shared/videos_controller.dart';

class InitialBinding implements Bindings {
  @override
  void dependencies() {
    Get.put(AuthController());
    Get.put(VideosController());
  }
}

void main() async {
  WidgetsFlutterBinding.ensureInitialized();
  await SystemChrome.setPreferredOrientations([
    DeviceOrientation.portraitUp,
    DeviceOrientation.portraitDown,
  ]);

  runApp(
    GetMaterialApp(
      /* smartManagement: SmartManagement.onlyBuilder, */
      initialBinding: InitialBinding(),
      debugShowCheckedModeBanner: false,
      defaultTransition: Transition.native,
      transitionDuration: Duration(milliseconds: 200),
      initialRoute: "/",
      getPages: AppRoutes.routes,
      title: 'ChallengeMe',
      theme: ThemeData(
        scaffoldBackgroundColor: Colors.white,
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
    ),
  );
}

```

**Comment by jonataslaw** [comment]
Fix on 3.4.0

**Comment by Kypsis** [comment]
Fixed as advertised. Thanks 👍
